File: src/main/java/io/github/chaosdave34/kitpvp/abilities/impl/enderman/DragonFireballAbility.java
package io.github.chaosdave34.kitpvp.abilities.impl.enderman;

import io.github.chaosdave34.kitpvp.KitPvp;
import io.github.chaosdave34.kitpvp.abilities.Ability;

import net.kyori.adventure.text.Component;
import org.bukkit.Location;
import org.bukkit.entity.AreaEffectCloud;
import org.bukkit.entity.DragonFireball;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.entity.CreatureSpawnEvent;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.metadata.FixedMetadataValue;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class DragonFireballAbility extends Ability {

    public DragonFireballAbility() {
        super("dragon_fireball", "Dragon Fireball", Type.RIGHT_CLICK, 20);
    }

    @Override
    public @NotNull List<Component> getDescription() {
        return createSimpleDescription("Shoots a fireball.");
    }

    @Override
    public boolean onAbility(Player p) {
        Location loc = p.getEyeLocation();
        p.getWorld().spawnEntity(loc.add(loc.getDirection().normalize()), EntityType.DRAGON_FIREBALL, CreatureSpawnEvent.SpawnReason.CUSTOM, (entity) -> {
            DragonFireball fireball = (DragonFireball) entity;
            fireball.setDirection(p.getLocation().getDirection().multiply(1.5));
            fireball.setShooter(p);
            fireball.setMetadata("ability", new FixedMetadataValue(KitPvp.INSTANCE, getId()));
        });
        return true;
    }

    // Todo: fix this to just work for fireballs from this ability and not every area effect cloud
    @EventHandler
    public void onDamage(EntityDamageByEntityEvent e) {
        if (e.getDamager() instanceof AreaEffectCloud areaEffectCloud) {
            if (areaEffectCloud.getSource() instanceof Player) {
                if (e.getEntity().getUniqueId().equals(areaEffectCloud.getOwnerUniqueId())) {
                    e.setCancelled(true);
                }
            }
        }
    }
}

File: src/main/kotlin/io/github/chaosdave34/kitpvp/abilities/impl/enderman/DragonFireballAbility.kt
package io.github.chaosdave34.kitpvp.abilities.impl.enderman

import io.github.chaosdave34.kitpvp.KitPvp
import io.github.chaosdave34.kitpvp.abilities.Ability
import net.kyori.adventure.text.Component
import org.bukkit.entity.AreaEffectCloud
import org.bukkit.entity.DragonFireball
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.metadata.FixedMetadataValue

class DragonFireballAbility : Ability("dragon_fireball", "Dragon Fireball", Type.RIGHT_CLICK, 20) {
    override fun getDescription(): List<Component> = createSimpleDescription("Shoots a dragon fireball.")

    override fun onAbility(player: Player): Boolean {
        val loc = player.eyeLocation
        player.world.spawn(loc.add(loc.direction.normalize()), DragonFireball::class.java) { fireball ->
            fireball.direction = player.location.direction.multiply(1.5)
            fireball.shooter = player
            fireball.setMetadata("ability", FixedMetadataValue(KitPvp.INSTANCE, id))
        }
        return true
    }

    // Todo: fix this to just work for fireballs from this ability and not every area effect cloud
    @EventHandler
    fun onDamage(event: EntityDamageByEntityEvent) {
        val areaEffectCloud = event.damager
        if (areaEffectCloud is AreaEffectCloud) {
            if (areaEffectCloud.source is Player) {
                if (event.entity.uniqueId == areaEffectCloud.ownerUniqueId) {
                    event.isCancelled = true
                }
            }
        }
    }
}